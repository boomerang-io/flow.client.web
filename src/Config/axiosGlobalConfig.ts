// Setup global config for Axios requests so they work with ICP
import axios from "axios";
axios.defaults.withCredentials = true;
