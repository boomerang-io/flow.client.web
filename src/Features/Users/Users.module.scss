.content {
  overflow: auto;
}

.tableRow {
  cursor: pointer;
}
