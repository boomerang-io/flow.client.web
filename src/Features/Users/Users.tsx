import React from "react";
import { Helmet } from "react-helmet";
import { useQuery } from "Hooks";
import { useHistory, useLocation, Route, Switch } from "react-router-dom";
import { Box } from "reflexbox";
import { DataTable, DataTableSkeleton, Pagination, Search } from "@carbon/react";
import {
  ErrorMessage,
  FeatureHeader as Header,
  FeatureHeaderTitle as HeaderTitle,
  FeatureHeaderSubtitle as HeaderSubtitle,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { isAccessibleKeyboardEvent } from "@boomerang-io/utils";
import EmptyState from "Components/EmptyState";
import UserDetailed from "Features/UserDetailed";
import debounce from "lodash/debounce";
import moment from "moment";
import queryString from "query-string";
import { CREATED_DATE_FORMAT, SortDirection } from "Constants";
import { AppPath, appLink } from "Config/appConfig";
import { serviceUrl } from "Config/servicesConfig";
import { FlowUser, PaginatedResponse } from "Types";
import styles from "./Users.module.scss";

const DEFAULT_ORDER = SortDirection.Desc;
const DEFAULT_PAGE = 0;
const DEFAULT_SIZE = 10;
const DEFAULT_SORT = "name";
const PAGE_SIZES = [DEFAULT_SIZE, 20, 50, 100];

const UsersContainer: React.FC = () => {
  return (
    <Switch>
      <Route path={AppPath.User}>
        <UserDetailed />
      </Route>
      <Route path={AppPath.UserList}>
        <UserList />
      </Route>
    </Switch>
  );
};

interface FeatureLayoutProps {
  handleSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const FeatureLayout: React.FC<FeatureLayoutProps> = ({ children, handleSearchChange }) => {
  return (
    <>
      <Helmet>
        <title>Users</title>
      </Helmet>
      <Header
        includeBorder={false}
        header={
          <>
            <HeaderTitle style={{ margin: "0" }}>Users</HeaderTitle>
            <HeaderSubtitle>View and manage Flow users</HeaderSubtitle>
          </>
        }
      />
      <Box p="2rem" className={styles.content}>
        <>
          <Box mb="1rem" maxWidth="20rem">
            <Search id="flow-users" labelText="Search users" placeholder="Search users" onChange={handleSearchChange} />
          </Box>
          {children}
        </>
      </Box>
    </>
  );
};

const UserList: React.FC = () => {
  const history = useHistory();
  const location = useLocation();

  const usersQuery = useQuery(serviceUrl.getManageUsers({ query: location.search }));

  function handleNavigateToUser(userId: string) {
    history.push(appLink.user({ userId }));
  }

  /**
   * Function that updates url search history to persist state
   * @param {object} query - all of the query params
   *
   */
  function updateHistorySearch({
    order = DEFAULT_ORDER,
    page = DEFAULT_PAGE,
    size = DEFAULT_SIZE,
    sort = DEFAULT_SORT,
    ...props
  }) {
    const queryStr = `?${queryString.stringify({ order, page, size, sort, ...props })}`;
    history.push({ search: queryStr });
  }
  // eslint-disable-next-line
  const debouncedSearch = React.useCallback(
    debounce((query: string) => {
      updateHistorySearch({ query, page: 0 });
    }, 300),
    []
  );

  function handleSearchChange(e: React.ChangeEvent<HTMLInputElement>) {
    const query = e.target.value;
    debouncedSearch(query);
  }

  function handlePaginationChange(pagination: { page: number; pageSize: number }) {
    updateHistorySearch({
      ...queryString.parse(location.search),
      page: pagination.page - 1, // We have to decrement by one to offset the table pagination adjustment
      size: pagination.pageSize,
    });
  }

  function handleSort(e: React.SyntheticEvent, sort: { sortHeaderKey: string }) {
    const { property, direction } = usersQuery.data.sort[0];
    let order = SortDirection.Asc;

    if (sort.sortHeaderKey === property && direction === SortDirection.Asc) {
      order = SortDirection.Desc;
    }

    updateHistorySearch({ ...queryString.parse(location.search), order, sort: sort.sortHeaderKey });
  }

  if (usersQuery.isLoading) {
    return (
      <FeatureLayout handleSearchChange={handleSearchChange}>
        <DataTableSkeleton />
      </FeatureLayout>
    );
  }

  if (usersQuery.isError) {
    return (
      <FeatureLayout handleSearchChange={handleSearchChange}>
        <ErrorMessage />
      </FeatureLayout>
    );
  }
  return (
    <FeatureLayout handleSearchChange={handleSearchChange}>
      <UsersTable
        handleNavigateToUser={handleNavigateToUser}
        handlePaginationChange={handlePaginationChange}
        handleSort={handleSort}
        usersData={usersQuery.data}
      />
    </FeatureLayout>
  );
};

const TableHeaderKey = {
  Name: "name",
  Email: "email",
  Type: "type",
  Created: "firstLoginDate",
  LastLogin: "lastLoginDate",
  Status: "status",
};

const headers = [
  {
    header: "Name",
    key: TableHeaderKey.Name,
    sortable: true,
  },
  {
    header: "Email",
    key: TableHeaderKey.Email,
    sortable: true,
  },
  {
    header: "Type",
    key: TableHeaderKey.Type,
    sortable: true,
  },
  {
    header: "Created",
    key: TableHeaderKey.Created,
    sortable: true,
  },
  {
    header: "Last Login",
    key: TableHeaderKey.LastLogin,
    sortable: true,
  },
  {
    header: "Status",
    key: TableHeaderKey.Status,
    sortable: true,
  },
];

interface UsersTableProps {
  handleNavigateToUser: (userId: string) => void;
  handlePaginationChange: (pagination: { page: number; pageSize: number }) => void;
  handleSort: (e: React.SyntheticEvent, sort: { sortHeaderKey: string }) => void;
  usersData: PaginatedResponse<FlowUser>;
}

const UsersTable: React.FC<UsersTableProps> = ({
  handleNavigateToUser,
  handlePaginationChange,
  handleSort,
  usersData,
}) => {
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;
  const { number: page, sort, totalElements, totalPages, records } = usersData;

  return records?.length > 0 ? (
    <>
      <DataTable
        rows={records}
        headers={headers}
        render={({ rows, headers, getHeaderProps }: any) => (
          <TableContainer>
            <Table isSortable>
              <TableHead>
                <TableRow>
                  {headers.map((header: any) => (
                    <TableHeader
                      id={header.key}
                      {...getHeaderProps({
                        header,
                        isSortable: header.sortable,
                        onClick: handleSort,
                      })}
                      isSortHeader={sort[0].property === header.key}
                      sortDirection={sort[0].direction}
                    >
                      {header.header}
                    </TableHeader>
                  ))}
                </TableRow>
              </TableHead>
              <TableBody>
                {rows.map((row: any) => (
                  <TableRow
                    className={styles.tableRow}
                    key={row.id}
                    onClick={() => handleNavigateToUser(row.id)}
                    onKeyDown={(e: React.SyntheticEvent) =>
                      isAccessibleKeyboardEvent(e) && handleNavigateToUser(row.id)
                    }
                    tabIndex={-1}
                  >
                    {row.cells.map((cell: any) => {
                      if (
                        cell.info.header === TableHeaderKey.Created ||
                        cell.info.header === TableHeaderKey.LastLogin
                      ) {
                        return <TableCell key={cell.id}>{moment(cell.value).format(CREATED_DATE_FORMAT)}</TableCell>;
                      }

                      if (Array.isArray(cell.value)) {
                        return <TableCell key={cell.id}>{cell.value.length}</TableCell>;
                      }

                      return <TableCell key={cell.id}>{cell.value}</TableCell>;
                    })}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        )}
      />
      <Pagination
        onChange={handlePaginationChange}
        page={page + 1}
        pageSize={totalPages}
        pageSizes={PAGE_SIZES}
        totalItems={totalElements}
      />
    </>
  ) : (
    <EmptyState message="No users found" />
  );
};

export default UsersContainer;
