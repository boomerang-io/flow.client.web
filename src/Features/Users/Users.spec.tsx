import React from "react";
import Users from ".";
import { Route } from "react-router-dom";
import { waitFor, screen, fireEvent } from "@testing-library/react";
import { queryCaches } from "react-query";
import { AppPath, appLink } from "Config/appConfig";
import { startApiServer } from "ApiServer";

let server;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
  queryCaches.forEach((queryCache) => queryCache.clear());
});

describe("Users --- Snapshot Test", () => {
  it("Capturing Snapshot of Users", async () => {
    const { baseElement, findByText } = rtlContextRouterRender(
      <Route path={AppPath.UserList}>
        <Users />
      </Route>,
      { route: appLink.userList() }
    );
    await findByText("Tim Bula");
    expect(baseElement).toMatchSnapshot();
    await waitFor(() => {});
  });
});

describe("Users --- RTL", () => {
  test("Change user role", async () => {
    rtlContextRouterRender(
      <Route path={AppPath.UserList}>
        <Users />
      </Route>,
      { route: appLink.userList() }
    );
    await screen.findByText(/^View and manage Flow users$/i);
    const overflowMenuButtons = await screen.findAllByTestId(/^user-menu$/i);

    fireEvent.click(overflowMenuButtons[0]);
    fireEvent.click(await screen.findByText(/^Change role$/i));
    expect(screen.getByText(/Admins can do more things/i)).toBeInTheDocument();
    expect(screen.getByText(/^Submit$/i)).toBeDisabled();
    fireEvent.click(await screen.findByText(/^User$/i));
    expect(screen.getByText(/^Submit$/i)).toBeEnabled();
    fireEvent.click(await screen.findByText(/^Submit$/i));
  });

  test("View user details", async () => {
    rtlContextRouterRender(
      <Route path={AppPath.UserList}>
        <Users />
      </Route>,
      { route: appLink.userList() }
    );
    await screen.findByText(/^View and manage Flow users$/i);
    const overflowMenuButtons = await screen.findAllByTestId(/^user-menu$/i);

    fireEvent.click(overflowMenuButtons[0]);
    fireEvent.click(await screen.findByText(/^View details$/i));
    expect(screen.getByText(/^Last Login$/i)).toBeInTheDocument();
  });
});
