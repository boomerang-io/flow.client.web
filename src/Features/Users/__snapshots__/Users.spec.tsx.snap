// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users --- Snapshot Test Capturing Snapshot of Users 1`] = `
<body
  id="app"
>
  <div>
    <header
      class="bx--bmrg-feature-header"
    >
      <section
        class="bx--bmrg-feature-header__content"
      >
        <div
          class="bx--bmrg-feature-header__header"
        >
          <h1
            class="bx--bmrg-feature-header-title"
            style="margin: 0px;"
          >
            Users
          </h1>
          <p
            class="bx--bmrg-feature-header-title"
          >
            View and manage Flow users
          </p>
        </div>
      </section>
    </header>
    <div
      class="css-1aaedcz"
    >
      <div
        class="css-2o893x"
      >
        <div
          aria-labelledby="flow-users-search"
          class="bx--search bx--search--xl"
          role="search"
        >
          <svg
            aria-hidden="true"
            class="bx--search-magnifier"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5	C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
            />
          </svg>
          <label
            class="bx--label"
            for="flow-users"
            id="flow-users-search"
          >
            Search users
          </label>
          <input
            autocomplete="off"
            class="bx--search-input"
            id="flow-users"
            placeholder="Search users"
            role="searchbox"
            type="text"
          />
          <button
            aria-label="Clear search input"
            class="bx--search-close bx--search-close--hidden"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="bx--data-table-container"
      >
        <div
          class="bx--data-table-content"
        >
          <table
            class="bx--data-table bx--data-table--sort bx--data-table--no-border"
          >
            <thead>
              <tr>
                <th
                  aria-sort="none"
                  scope="col"
                >
                  <button
                    class="bx--table-sort"
                    id="name"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Name
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  scope="col"
                >
                  <button
                    class="bx--table-sort"
                    id="email"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Email
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  scope="col"
                >
                  <button
                    class="bx--table-sort"
                    id="type"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Type
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="ascending"
                  scope="col"
                >
                  <button
                    class="bx--table-sort bx--table-sort--active"
                    id="firstLoginDate"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Created
                      </div>
                      <svg
                        aria-label="Sort rows by this header in descending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in descending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  scope="col"
                >
                  <button
                    class="bx--table-sort"
                    id="status"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Status
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  id="action"
                  scope="col"
                />
              </tr>
            </thead>
            <tbody
              aria-live="polite"
            >
              <tr>
                <td>
                  Tim Bula
                </td>
                <td>
                  trbula@us.ibm.com
                </td>
                <td>
                  admin
                </td>
                <td>
                  January 01, 2020
                </td>
                <td>
                  active
                </td>
                <td>
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Menu"
                    class="bx--overflow-menu"
                    data-testid="user-menu"
                    type="button"
                  >
                    <svg
                      aria-label="open and close list of options"
                      class="bx--overflow-menu__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="16"
                        cy="8"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="16"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="24"
                        r="2"
                      />
                      <title>
                        open and close list of options
                      </title>
                    </svg>
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  Marcus Roy
                </td>
                <td>
                  mdroy@us.ibm.com
                </td>
                <td>
                  admin
                </td>
                <td>
                  January 01, 2020
                </td>
                <td>
                  active
                </td>
                <td>
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Menu"
                    class="bx--overflow-menu"
                    data-testid="user-menu"
                    type="button"
                  >
                    <svg
                      aria-label="open and close list of options"
                      class="bx--overflow-menu__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="16"
                        cy="8"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="16"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="24"
                        r="2"
                      />
                      <title>
                        open and close list of options
                      </title>
                    </svg>
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  Test User
                </td>
                <td>
                  test@ibm.com
                </td>
                <td>
                  admin
                </td>
                <td>
                  January 01, 2020
                </td>
                <td>
                  active
                </td>
                <td>
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Menu"
                    class="bx--overflow-menu"
                    data-testid="user-menu"
                    type="button"
                  >
                    <svg
                      aria-label="open and close list of options"
                      class="bx--overflow-menu__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="16"
                        cy="8"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="16"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="24"
                        r="2"
                      />
                      <title>
                        open and close list of options
                      </title>
                    </svg>
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="bx--pagination"
      >
        <div
          class="bx--pagination__left"
        >
          <label
            class="bx--pagination__text"
            for="bx-pagination-select-1"
            id="bx-pagination-select-1-count-label"
          >
            Items per page:
          </label>
          <div
            class="bx--form-item"
          >
            <div
              class="bx--select bx--select--inline bx--select__item-count"
            >
              <div
                class="bx--select-input--inline__wrapper"
              >
                <div
                  class="bx--select-input__wrapper"
                >
                  <select
                    class="bx--select-input"
                    id="bx-pagination-select-1"
                  >
                    <option
                      class="bx--select-option"
                      value="10"
                    >
                      10
                    </option>
                    <option
                      class="bx--select-option"
                      value="20"
                    >
                      20
                    </option>
                    <option
                      class="bx--select-option"
                      value="50"
                    >
                      50
                    </option>
                    <option
                      class="bx--select-option"
                      value="100"
                    >
                      100
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="bx--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="bx--pagination__text bx--pagination__items-count"
          >
            1–2 of 2 items
          </span>
        </div>
        <div
          class="bx--pagination__right"
        >
          <div
            class="bx--form-item"
          >
            <div
              class="bx--select bx--select--inline bx--select__page-number"
            >
              <label
                class="bx--label bx--visually-hidden"
                for="bx-pagination-select-1-right"
              >
                Page number, of 1 pages
              </label>
              <div
                class="bx--select-input--inline__wrapper"
              >
                <div
                  class="bx--select-input__wrapper"
                >
                  <select
                    class="bx--select-input"
                    id="bx-pagination-select-1-right"
                  >
                    <option
                      class="bx--select-option"
                      value="1"
                    >
                      1
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="bx--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="bx--pagination__text"
          >
            of 1 page
          </span>
          <div
            class="bx--pagination__control-buttons"
          >
            <button
              class="bx--pagination__button bx--pagination__button--backward bx--pagination__button--no-index bx--btn bx--btn--ghost bx--btn--disabled bx--btn--icon-only bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--top bx--tooltip--align-center"
              disabled=""
              tabindex="0"
              type="button"
            >
              <span
                class="bx--assistive-text"
              >
                Previous page
              </span>
              <svg
                aria-hidden="true"
                aria-label="Previous page"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 24L10 16 20 8z"
                />
              </svg>
            </button>
            <button
              class="bx--pagination__button bx--pagination__button--forward bx--pagination__button--no-index bx--btn bx--btn--ghost bx--btn--disabled bx--btn--icon-only bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--top bx--tooltip--align-end"
              disabled=""
              tabindex="0"
              type="button"
            >
              <span
                class="bx--assistive-text"
              >
                Next page
              </span>
              <svg
                aria-hidden="true"
                aria-label="Next page"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8L22 16 12 24z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
